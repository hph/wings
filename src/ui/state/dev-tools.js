import React from 'react';
import { createDevTools } from 'redux-devtools';
import LogMonitor from 'redux-devtools-log-monitor';
import DockMonitor from 'redux-devtools-dock-monitor';
import Dispatcher from 'redux-devtools-dispatch';
import MultipleMonitors from 'redux-devtools-multiple-monitors';

export default createDevTools(
  <DockMonitor
    toggleVisibilityKey="ctrl-h"
    changePositionKey="ctrl-q"
    changeMonitorKey="ctrl-m"
    defaultIsVisible={false}
  >
    <MultipleMonitors>
      <LogMonitor theme="solarized" />
      <Dispatcher theme="solarized" />
    </MultipleMonitors>
  </DockMonitor> // eslint-disable-line comma-dangle
);
