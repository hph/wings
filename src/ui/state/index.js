export * as actions from 'ui/state/actions';
export * as selectors from 'ui/state/selectors';
export * as types from 'ui/state/types';
export configureStore from 'ui/state/store/shared';
export middleware from 'ui/state/middleware';
export reducer from 'ui/state/reducer';
