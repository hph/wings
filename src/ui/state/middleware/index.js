import { applyMiddleware } from 'redux';

import userInput from 'ui/state/middleware/user-input';
import cursorPosition from 'ui/state/middleware/cursor-position';
import commandBar from 'ui/state/middleware/command-bar';

export default applyMiddleware(
  userInput,
  cursorPosition,
  commandBar,
);
