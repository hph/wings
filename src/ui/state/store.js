import { applyMiddleware, createStore, compose } from 'redux';

import {
  DevTools,
  reducer,
  createUserInputMiddleware,
} from 'ui/state';


export default function configureStore (initialState) {
  let enhancer = applyMiddleware(createUserInputMiddleware());

  if (process.env.NODE_ENV === 'development') {
    enhancer = compose(enhancer, DevTools.instrument({ maxAge: 30 }));
  }

  return createStore(reducer, initialState, enhancer);
}
