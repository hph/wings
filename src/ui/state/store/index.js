const file = process.env.NODE_ENV === 'production'
  ? 'production'
  : 'development'; // Also use this store in the test environment.
// eslint-disable-next-line import/no-dynamic-require
module.exports = require(`./${ file }`).default;
