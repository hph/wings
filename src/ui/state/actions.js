import { types } from 'ui/state';

let nextPaneId = 0;
export function createPane(filename, text) {
  nextPaneId += 1;
  return {
    type: types.CREATE_PANE,
    id: nextPaneId,
    filename,
    text,
  };
}

export function updatePane(id, options) {
  return {
    type: types.UPDATE_PANE,
    id,
    ...options,
  };
}

export function destroyPane(id) {
  return {
    type: types.DESTROY_PANE,
    id,
  };
}

export function updateConfig(options) {
  return {
    type: types.UPDATE_CONFIG,
    ...options,
  };
}

export function updateCommand(options) {
  return {
    type: types.UPDATE_COMMAND,
    ...options,
  };
}

export function execCommand(options) {
  return {
    type: types.EXEC_COMMAND,
    ...options,
  };
}

export function toggleTreeView() {
  return {
    type: types.TOGGLE_TREE_VIEW,
  };
}

export function userInputFocus(focus) {
  return {
    type: focus ? types.USER_INPUT_FOCUS : types.USER_INPUT_UNFOCUS,
  };
}
