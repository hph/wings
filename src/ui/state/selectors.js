import _ from 'lodash';
import { createSelector } from 'reselect';

import { computeFontDimensions } from 'ui/utils';

export const currentPane = createSelector(
  state => state.panes,
  state => state.config.currentPaneId,
  (panes, currentPaneId) => currentPaneId
    ? _.find(panes, ({ id }) => id === currentPaneId)
    : panes[0],
);

export const paneById = createSelector(
  state => state.panes,
  (state, props) => props.paneId,
  (panes, paneId) => _.find(panes, ({ id }) => id === paneId),
);

export const charSizes = createSelector(
  state => state.config.theme,
  theme => {
    const { width, height } = computeFontDimensions(theme);
    return {
      charHeight: height,
      charWidth: width,
    };
  },
);
