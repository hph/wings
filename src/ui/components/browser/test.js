import React from 'react';
import renderer from 'react-test-renderer';

import { userInputFocus } from 'ui/state/actions';
import { Browser, withScheme } from './index';

function createNodeMockCreator (methods = {}) {
  return (element) => {
    if (element.type === 'input') {
      return {
        select: () => {},
        ...methods,
      };
    }

    if (element.type === 'webview') {
      return {
        addEventListener: () => {},
        reload: () => {},
        ...methods,
      };
    }

    return null;
  };
}

const defaultProps = {
  dispatch: () => {},
};

const createTree = (props = defaultProps, mockMethods = {}) => {
  const createNodeMock = createNodeMockCreator(mockMethods);
  const component = renderer.create(<Browser {...props} />, { createNodeMock });
  return {
    component,
    tree: component.toJSON(),
  };
};

const createSnapshot = (props = defaultProps, mockMethods = {}) => {
  const { tree } = createTree(props, mockMethods);
  expect(tree).toMatchSnapshot();
};

describe('Browser', () => {
  it('renders a default browser view', () => {
    createSnapshot();
  });

  it('selects the navigation bar on mount', done => {
    const select = jest.fn();
    createTree(defaultProps, { select });

    setTimeout(() => {
      expect(select.mock.calls.length).toBe(1);
      done();
    });
  });

  it('sets up event listeners on mount', () => {
    const addEventListener = jest.fn();
    createTree(defaultProps, { addEventListener });

    expect(addEventListener.mock.calls.length).toBe(3);
    expect(addEventListener.mock.calls[0][0]).toEqual('will-navigate');
    expect(addEventListener.mock.calls[1][0]).toEqual('did-get-redirect-request');
    expect(addEventListener.mock.calls[2][0]).toEqual('did-fail-load');
  });

  it('dispatches an action to stop user input focus on click', () => {
    const dispatch = jest.fn();
    const { tree } = createTree({ ...defaultProps, dispatch });
    tree.children[0].props.onClick();

    expect(dispatch.mock.calls.length).toBe(1);
    expect(dispatch.mock.calls[0][0]).toEqual(userInputFocus(false));
  });

  it('updates the value in the navbar as the user types', () => {
    const { component, tree } = createTree();
    expect(tree).toMatchSnapshot();
    tree.children[0].props.onChange({ target: { value: 'example.com' } });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('navigates to the URL in the navbar on an Enter keypress', () => {
    const reload = jest.fn();
    const { component, tree } = createTree(defaultProps, { reload });
    tree.children[0].props.onChange({ target: { value: 'example.com' } });
    tree.children[0].props.onKeyDown({ key: 'Enter' });
    expect(component.toJSON()).toMatchSnapshot();
    expect(reload.mock.calls.length).toBe(0);

    // Browser calls reload when the URL does not change on an enter keypress.
    tree.children[0].props.onKeyDown({ key: 'Enter' });
    expect(reload.mock.calls.length).toBe(1);

    // Non-enter key presses don't do anything.
    tree.children[0].props.onKeyDown({ key: 'a' });
    expect(reload.mock.calls.length).toBe(1);
  });

  it('shows an error on a webview load error', () => {
    let onDidFailLoad;
    const addEventListener = jest.fn((event, callback) => {
      if (event === 'did-fail-load') {
        onDidFailLoad = callback;
      }
    });
    const { component } = createTree(defaultProps, { addEventListener });
    onDidFailLoad({ errorDescription: 'oh no!' });

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('withScheme function prepends an HTTP scheme to a string', () => {
    // Defaults to using http over https, unless provided.
    expect(withScheme('foo.com')).toEqual('http://foo.com');
    expect(withScheme('http://foo.com')).toEqual('http://foo.com');
    expect(withScheme('https://foo.com')).toEqual('https://foo.com');
  });

  describe('mapStateToProps', () => {
    it('returns an empty object', () => {
      // The store provides the dispatch prop.
      expect(Browser.mapStateToProps({})).toEqual({});
    });
  });

  describe('onWillNavigate', () => {
    it('updates the navLocation', () => {
      const browser = new Browser();
      browser.setState = jest.fn();
      browser.onWillNavigate({ url: 'example.com' });
      expect(browser.setState).toHaveBeenCalledWith({
        navLocation: 'example.com',
      });
    });
  });

  describe('onDidGetRedirectRequest', () => {
    it('updates the navLocation', () => {
      const browser = new Browser();
      browser.setState = jest.fn();

      browser.onDidGetRedirectRequest({ newURL: '', isMainFrame: false });
      expect(browser.setState).not.toHaveBeenCalled();

      browser.onDidGetRedirectRequest({ newURL: '', isMainFrame: true });
      expect(browser.setState).not.toHaveBeenCalled();

      browser.onDidGetRedirectRequest({ newURL: 'example.com', isMainFrame: false });
      expect(browser.setState).not.toHaveBeenCalled();

      browser.onDidGetRedirectRequest({ newURL: 'example.com', isMainFrame: true });
      expect(browser.setState).toHaveBeenCalledWith({
        navLocation: 'example.com',
      });
    });
  });

  describe('onDidFailLoad', () => {
    it('Browser onDidFailLoad updates the state with an error message', () => {
      const browser = new Browser();
      browser.setState = jest.fn();
      browser.onDidFailLoad({ errorDescription: 'Not found!' });
      expect(browser.setState).toHaveBeenCalledWith({
        error: 'Not found!',
      });
    });
  });
});
