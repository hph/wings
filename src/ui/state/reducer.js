import _ from 'lodash';
import { combineReducers } from 'redux';

import fixedKeys from 'ui/fixed-keys';
import * as types from 'ui/state/types';

export function configReducer(state = {}, action) {
  const { type, ...values } = action;
  switch (type) {
    case types.UPDATE_CONFIG: {
      return {
        ...state,
        ...values,
      };
    }

    case types.TOGGLE_TREE_VIEW: {
      return {
        ...state,
        isTreeViewVisible: !state.isTreeViewVisible,
      };
    }

    default:
      return state;
  }
}

export function panesReducer(state = [], action) {
  const { type, ...values } = action;
  switch (type) {
    case types.CREATE_PANE: {
      const lines = (action.text || '').split('\n').slice(0, -1);
      return [
        ...state,
        _.defaults(
          { ..._.omit(values, ['text']) },
          {
            lines: _.isEmpty(lines) ? [''] : lines,
            column: 0,
            row: 0,
            firstVisibleRow: 0,
            firstVisibleColumn: 0,
            width: window.screen.width,
            height: window.screen.height,
          },
        ),
      ];
    }

    case types.UPDATE_PANE: {
      return _.map(state, pane => {
        if (pane.id === action.id) {
          return _.defaults({ ...values }, pane);
        }
        return pane;
      });
    }

    case types.DESTROY_PANE: {
      return _.reject(state, ({ id }) => id === values.id);
    }

    default:
      return state;
  }
}

export function commandReducer(state = '', action) {
  const { type, value } = action;
  switch (type) {
    case types.UPDATE_COMMAND: {
      if (fixedKeys.has(value)) {
        if (value === 'Space') {
          return `${state} `;
        } else if (value === 'Backspace') {
          return state.substring(0, state.length - 1);
        } else if (value === 'Enter') {
          return '';
        }
        return state;
      }
      return state + value;
    }

    default:
      return state;
  }
}

export default combineReducers({
  config: configReducer,
  panes: panesReducer,
  command: commandReducer,
});
