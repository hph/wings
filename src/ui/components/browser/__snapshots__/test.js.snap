// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Browser navigates to the URL in the navbar on an Enter keypress 1`] = `
<div
  className="root"
>
  <input
    autoFocus={true}
    className="navigation"
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="http://example.com"
  />
  
  <webview
    className="webview"
    src="http://example.com"
  />
</div>
`;

exports[`Browser renders a default browser view 1`] = `
<div
  className="root"
>
  <input
    autoFocus={true}
    className="navigation"
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="http://google.com"
  />
  
  <webview
    className="webview"
    src=""
  />
</div>
`;

exports[`Browser shows an error on a webview load error 1`] = `
<div
  className="root"
>
  <input
    autoFocus={true}
    className="navigation"
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="http://google.com"
  />
  <div
    className="errorContainer"
  >
    <div>
      Error: 
      oh no!
    </div>
  </div>
  <webview
    className="webview"
    src=""
  />
</div>
`;

exports[`Browser updates the value in the navbar as the user types 1`] = `
<div
  className="root"
>
  <input
    autoFocus={true}
    className="navigation"
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="http://google.com"
  />
  
  <webview
    className="webview"
    src=""
  />
</div>
`;

exports[`Browser updates the value in the navbar as the user types 2`] = `
<div
  className="root"
>
  <input
    autoFocus={true}
    className="navigation"
    onChange={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="text"
    value="example.com"
  />
  
  <webview
    className="webview"
    src=""
  />
</div>
`;
